casper/src/main/scala/coop/rchain/casper/merging/BlockHashDagMerger.scala

```diff
@@ -0,0 +1,192 @@
+package coop.rchain.casper.merging
+
+import cats.effect.Concurrent
+import com.google.protobuf.ByteString
+import coop.rchain.blockstorage.dag.Message
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.Validator.Validator
+import coop.rchain.models.syntax._
+import coop.rchain.rholang.interpreter.RhoRuntime.RhoHistoryRepository
+import coop.rchain.rholang.interpreter.merging.RholangMergingLogic
+import coop.rchain.rspace.HotStoreTrieAction
+import coop.rchain.rspace.hashing.Blake2b256Hash
+import coop.rchain.rspace.merger.{ChannelChange, StateChange, StateChangeMerger}
+import coop.rchain.rspace.merger.EventLogMergingLogic.NumberChannelsDiff
+import coop.rchain.sdk.dag.merging.DagMergingLogic
+import coop.rchain.sdk.dag.merging.DagMergingLogic._
+import coop.rchain.shared.{Log, Stopwatch}
+import coop.rchain.shared.syntax._
+import scodec.bits.ByteVector
+import cats.syntax.all._
+import coop.rchain.rholang.syntax._
+import coop.rchain.rspace.syntax._
+
+object BlockHashDagMerger {
+
+  final case object UnableToMergeInBootstrapMode extends Exception("Node is in bootstrap mode.")
+
+  def merge[F[_]: Concurrent: Log](
+      tips: Set[BlockHash],
+      finalFringe: Set[BlockHash],
+      finalState: Blake2b256Hash,
+      merger: BlockHashDagMerger,
+      fringeStates: Map[Set[BlockHash], (Blake2b256Hash, Set[ByteString])],
+      historyRepository: RhoHistoryRepository[F],
+      blockIndex: BlockHash => F[Option[BlockIndex]],
```

Comment:
Why to accept optional result if it must be found?

Reply:
`UnableToMergeInBootstrapMode` exception is raised if some of calls return None. Do you want to check bootstrap in another place?

Reply:
I removed this in <https://github.com/rchain/rchain/pull/3742/commits/fb2384fde99efb5f1b679781cb4f48cde4689763>